pre_commit/languages/dotnet.py

```diff
@@ -0,0 +1,108 @@
+import contextlib
+import glob
+import os.path
+import shutil
+from typing import Generator
+from typing import Sequence
+from typing import Tuple
+
+import pre_commit.constants as C
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.envcontext import Var
+from pre_commit.hook import Hook
+from pre_commit.languages import helpers
+from pre_commit.prefix import Prefix
+from pre_commit.util import clean_path_on_failure
+from pre_commit.util import rmtree
+
+ENVIRONMENT_DIR = 'dotnetenv'
+BIN_DIR = 'bin'
+BUILD_CONFIG = 'Release'
+
+get_default_version = helpers.basic_get_default_version
+healthy = helpers.basic_healthy
+
+
+def get_env_patch(venv: str) -> PatchesT:
+
+    # https://github.com/dotnet/sdk/issues/9304
+    dotnet = shutil.which('dotnet')
+    assert dotnet, 'No dotnet executable found - please install one'
+    dotnet_root = os.path.dirname(os.path.realpath(dotnet))
+
+    return (
+        ('PATH', (os.path.join(venv, BIN_DIR), os.pathsep, Var('PATH'))),
+        ('DOTNET_ROOT', dotnet_root),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix) -> Generator[None, None, None]:
+    directory = helpers.environment_dir(ENVIRONMENT_DIR, C.DEFAULT)
+    envdir = prefix.path(directory)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+
+    helpers.assert_version_default('dotnet', version)
+    helpers.assert_no_additional_deps('dotnet', additional_dependencies)
+
+    envdir = helpers.environment_dir(ENVIRONMENT_DIR, version)
+    repodir = prefix.path(envdir)
+
+    with clean_path_on_failure(repodir):
+
+        build_dir = 'pre-commit-build'
+
+        # Build & pack nupkg file
+        helpers.run_setup_cmd(
+            prefix,
+            (
+                'dotnet', 'pack',
+                '--configuration', BUILD_CONFIG,
+                '--output', build_dir,
+            ),
+        )
+
+        # Determine the tool name
+        build_outputs = os.listdir(os.path.join(prefix.prefix_dir, build_dir))
+        assert (
+            len(build_outputs) == 1
+        ), f"Can't handle multiple build outputs. Got {build_outputs}"
+        tool_name = build_outputs[0].split('.')[0]
+
+        # Install to bin dir
+        helpers.run_setup_cmd(
+            prefix,
+            (
+                'dotnet', 'tool', 'install',
+                '--tool-path', os.path.join(repodir, BIN_DIR),
+                '--add-source', build_dir,
+                tool_name,
+            ),
+        )
+
+        # Cleanup
+        for d in glob.glob(f'{prefix.prefix_dir}/*'):
+            if d.endswith(envdir):
+                continue
+            if os.path.isfile(d):
+                os.remove(d)
+            else:
+                rmtree(d)
```

Comment:
could this be fixed by using a temporary directory to build and then just copy the products into the envdir?

Reply:
Yeah I'm not sure if/how this is done for other languages -- it seems like the hook repo is already cloned before `install_environment` is called so there's no chance by that point to clone/build in a temp dir.

I had a minimal version before which removed just the `bin`, `obj`, and `pre-commit-build` dirs since those are the ones we know will exist. I can revert to that if you'd prefer? Or we could skip any cleanup completely. The main reason I added this was that the `format` repo I used in the example above is chunky, and all we need after install is the `envdir`.

Reply:
I think `go` does a similar thing and build a temporary build directory and then copies the binaries out of it -- maybe that can be some inspiration?

Reply:
With `go` we seem to specify where the build artefacts will be placed. I'm not sure we can easily do here without doing bota a `dotnet build` cmd and then a `dotnet pack --no-build` and somehow finding the build output (which can be anywhere really depending on the project layout).

Could we do a `git clean` and exclude the final tool directory?

Reply:
a `git clean` would unfortunately preclude doing 1st class in the future so we should avoid that (multiple `language_version`s end up in the same clone directory)